doc/LanguageClient.txt

```diff
@@ -700,6 +674,117 @@ Default: []
 For a full list of the possible token types (name) see the LSP documentation:
 https://microsoft.github.io/language-server-protocol/specifications/specification-current/
 
+==============================================================================
+3. Settings                                           *LanguageClientSettings*
+
+Settings (aka "initialization options") for Language servers, as well as
+settings for some LanguageClient extensions (e.g. `rust-analyzer`) are specified
+in a single or in multiple `settings.json` files. Settings files locations
+are configured by `g:LanguageClient_settingsPath` variable.
+
+The initialization options found in the files are combined with
+the initialization options specified in the server command, if any.
+The former taking precedence over the latter.
+
+Note that the key under which the initialization options for each server lives
+matches the key under which the server expects it's configuration. This is
+important for servers that can request configuration via a
+`workspace/configuration` request.
+
+Previously, the initialization options lived under a `initializationOptions`
+key, which worked, but made answering that `workspace/configuration` request
+hard, since we couldn't really get the correct path to the setting the server
+requested. Since version 0.1.161 of this plugin, that key has been deprecated
+and you'll see a message saying that you should move off of it if your settings
+file includes an `initializationOptions` key.
+
+Example settings file content: >
+    {
+        "gopls": {
+          "usePlaceholders": true,
+          "local": "github.com/my/pkg",
+        },
+        "rust-analyzer": {
+          "inlayHints": {
+            "enable": true,
+            "chainingHints": true
+          },
+        },
+    }
+
+3.1 rust-analyzer     *LanguageClientSettings_rust-analyzer*
```

Comment:
>I think the confusion may arise from the fact that we are using those settings as well to enable some calls to the server and whatnot, so we are essentially using them as settings to our extensions as well, but that's purely for simplicity and to keep it transparent for the user.

Then how the user can learn about our extensions settings? Only by looking at the source code?
If you don't want to put this info in a separate `Settings` section, then suggest where it belongs better.

Reply:
I think I'm not communicating what I want to communicate correctly. I of course don't expect the user to look at the source code to figure out about the hints. The hints are a feature offered by rust-analyzer in this case, and they are documented there, we don't have any specific toggles for them because they are very specific to rust-analyzer, and we can use the initialization options provided by the user (in `settings.json`) as a config for our side of the feature anyways.

The initialization options, and therefore settings.json file, is a little bit of an obscure corner of the protocol precisely because each server "expects" different values in there. And I wouldn't want to go on the rabbit hole of adding each and every option for each and every server, that's up for each server to document.

Reply:
>  toggle displaying different kinds of inlay type hints,

This is controlled by the server initialization options. We merely react to the value of the setting by doing stuff client-side so that the setting that we passed on to the server makes sense.

>  customize prefixes for them

I'm honestly not sure where something like this would live in our config/docs, but I'm also not sure it's necessary (provided we get it right at some point), going a little bit on a tangent, I would honestly remove the customization of these until requested, we can of course keep it parameterized though, so that it's easier to customize the time for that eventually comes.

> Another solution would be to move these toggles to vim variables of the form g:LanguageClient_settings_*.

I mean, maybe we'll end up going to something like server specific setting on a dictionary under some sort of 
```
let g:LanguageClient_extensions = {}
let g:LanguageClient_extensions['rust-analyzer'] = { 'chainingHintsPrefix': ' - ' }
```
But I don't think there's a justification for that now, as the toggles we can control from the settings sent to the server (if the user enables them then it should be clear enough for us that they want us to display them), and the prefixes and separator I'm not convinced need customization right now.